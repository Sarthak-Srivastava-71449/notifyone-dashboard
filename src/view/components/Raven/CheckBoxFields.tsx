import { Checkbox } from 'antd';
import React from 'react';

interface CheckBoxFieldsProps {
  changeFunction: any;
  name: string;
  id: string;
  fields: string;
  checkedField: any;
  disabledField: any;
  isFieldChecked: boolean;
  skusCheck: boolean;
}

const CheckBoxFields: React.FC<CheckBoxFieldsProps> = (
  props: CheckBoxFieldsProps
) => {
  let checked = false;
  let disabled = false;
  if (props.checkedField && props.checkedField.indexOf(props.fields) > -1) {
    checked = true;
  }
  if (
    (props.disabledField && props.disabledField.indexOf(props.fields) > -1) ||
    props.isFieldChecked ||
    props.skusCheck
  ) {
    disabled = true;
  }
  return (
    <Checkbox
      id={props.name + props.id}
      name={props.name}
      className="regular-checkbox"
      value={props.fields}
      onChange={(e) => props.changeFunction(props.fields)}
      checked={checked}
      disabled={disabled}
    >
      {props.fields}
    </Checkbox>
  );
};

export default CheckBoxFields;
