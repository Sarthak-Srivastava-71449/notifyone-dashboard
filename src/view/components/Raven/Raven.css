/* Root Node css start */
.raven-wrapper {
  max-width: 90%;
  margin: 40px auto;
  bottom: 10px;
  position: relative;
  background: #eee;
  padding: 12px;
  border-radius: 6px;
  box-shadow: 7px 10px 10px 14px #fff;
}
.raven-wrapper
  .ant-collapse
  > .ant-collapse-item
  > .ant-collapse-header
  .ant-collapse-arrow {
  font-size: 14px !important;
  vertical-align: 2px !important;
}

.raven-wrapper .ant-collapse {
  border: 0 !important;
  background: #fff !important;
  margin: 12px 0 !important;
}

.raven-wrap {
  width: 75%;
}
.ant-collapse-item-disabled > .ant-collapse-header .ant-collapse-arrow {
  color: rgba(0, 0, 0, 0.25) !important;
}
.raven-wrapper .ant-collapse > .ant-collapse-item > .ant-collapse-header {
  padding: 12px 12px !important;
  align-items: center !important;
  box-shadow: 2px 2px 0px 0px rgb(0 0 0 / 8%);
  border-radius: 6px !important;
}

.ant-collapse-content > .ant-collapse-content-box {
  background: #eee !important;
}
.link {
  color: #1890ff;
}
link:hover {
  color: #40a9ff;
}
.ant-collapse {
  position: relative;
}
.node-link {
  width: 100%;
  position: absolute;
  top: 0px;
  left: 0;
}
.link.edit-option {
  display: flex;
  align-items: center;
}
.raven-wrapper
  .ant-collapse
  > .ant-collapse-item
  > .ant-collapse-header
  .ant-collapse-header-text {
  display: inline-block;
  inline-size: 40%;
  flex: none;
  margin-right: 20px;
  color: #222 !important;
  word-break: break-word;
  text-align: justify;
}

.condition-btn {
  display: flex;
  justify-content: center;
  font-size: 16px;
  height: 25px;
  align-items: center;
  box-shadow: 0 2px 2px 0 #fff;
  border-radius: 4px;
  width: 27px;
  background: #0000000d;
  border: 1px solid #0000000d;
}

.raven-wrapper
  .ant-collapse
  > .ant-collapse-item
  > .ant-collapse-header
  .ant-collapse-extra {
  height: 35px;
  width: 100%;
  margin: 0 auto;
  position: relative;
}
.ant-collapse-extra > .edit-option {
  display: flex;
  align-items: center;
}
.raven-wrapper .anticon.anticon-edit {
  padding: 0 8px;
}
.inner-item {
  position: absolute;
  width: 100%;
  left: 0px;
  top: 0px;
  display: flex;
  align-items: center;
}

.edit-node-btn,
.edit-link {
  margin-left: auto;
}

.raven-wrapper input.ant-input.input-number {
  width: 200px;
}

.raven-form-title {
  color: #1890ff;
}
/* Root Node css end */

/* Edit Action css start */
.wallet-wrapper {
  background-color: #fff;
  border-radius: 6px;
}

.wallet-wrapper .ant-select-selector {
  height: 34px !important;
}

.wallet-wrapper .ant-input {
  padding: 4px 5px !important;
}

.node-name {
  padding: 4px 5px !important;
}

.wallet-wrapper .ant-select-selector {
  padding: 0 7px !important;
}

.screen-test {
  border: 1px solid #a9a9a9;
  color: #a9a9a9;
  padding: 0 12px;
  margin-top: 8px;
}
.no-character > span {
  color: #a9a9a9;
}

.linknode-details-id {
  color: #a9a9a9;
}

.no-character .character {
  color: #a9a9a9;
  font-weight: 500;
}
/* Edit Action css end */

/* Edit Node css start */
.width-9 {
  width: 10%;
}
.width-91 {
  width: 91%;
}

.ant-collapse-content.ant-collapse-content-active
  .ant-collapse-header
  .edit-node-btn
  .parent-edit-btn {
  display: none !important;
}

.raven-wrapper
  .ant-collapse
  > .ant-collapse-item
  > .ant-collapse-header
  .child-edit-btn {
  display: none;
}

.ant-collapse-content.ant-collapse-content-active
  .ant-collapse-header
  .edit-node-btn
  .child-edit-btn {
  display: block;
}

.raven-wrapper .ant-checkbox-group {
  padding: 0 12px;
}
/* Edit Node css end */
