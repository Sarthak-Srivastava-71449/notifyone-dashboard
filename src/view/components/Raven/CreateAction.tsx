import React, { Fragment, useEffect, useReducer, useState } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { Select, Input } from 'antd';
import { Button } from '../Button';
import './Raven.css';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from 'src/store';
import {
  fetchRavenMetaData,
  fetchRavenTickets,
} from 'src/store/actions/ravenRootNodeEvents';
import {
  addNewNodeAction,
  fetchRavenNodeActions,
} from 'src/store/actions/ravenActionNodeEvents';
import { useNavigate } from 'react-router-dom';
import ravenConstants from 'src/common/constants/ravenConstants';
import { toast } from 'react-toastify';
import CheckBoxFields from './CheckBoxFields';
import {
  getCreateSuccess,
  getRavenMetaInfo,
  getTicketOption,
} from 'src/store/selectors/createAction';
import { getAccessToken } from 'src/store/selectors/accessToken';

const CreateAction: React.FC = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  /*
    Redux Selectors for fetching data
  */

  const ticketPriorityObjects: Array<any> = useSelector((state: RootState) =>
    getTicketOption(state)
  );
  const ravenMetaInfo = useSelector((state: RootState) =>
    getRavenMetaInfo(state)
  );
  const accessToken = useSelector((state: RootState) => getAccessToken(state));
  const createSuccess = useSelector((state: RootState) =>
    getCreateSuccess(state)
  );

  /*
   UseEffect for loading Raven Node Actions Data
  */

  useEffect(() => {
    if (createSuccess === true) {
      dispatch(fetchRavenNodeActions(accessToken, 200, 0));
      navigate('/communication/raven', { replace: true });
    }
  }, [createSuccess]);

  const [optionalFields, setOptionalFields] = useState<Array<string>>([]);
  const [requiredFields, setRequiredFields] = useState<Array<string>>([]);

  /*
   UseEffect for Fetching Raven Meta Data and Ticket Data
 */

  useEffect(() => {
    dispatch(fetchRavenMetaData(accessToken));
    dispatch(fetchRavenTickets(accessToken));
  }, []);

  const [isPreview, setIsPreview] = useState(false);

  const editActionInputState = {
    headerName: '',
    screenText: '',
    responseText:
      '<b>Thank You!</b><br>We have received your message and will be in touch over your registered email id as soon as possible.',
    type: '',
    ticketPriority: '',
  };

  /*
   React reducer for managing inputState
  */

  const editActionReducer = (state = editActionInputState, action: any) => {
    switch (action.type) {
      case 'INPUT_CHANGE':
        return { ...state, ...action.payload };
    }
  };
  const [editState, editDispatch] = useReducer(
    editActionReducer,
    editActionInputState
  );

  const updateInputField = (event: any, inputField: string) => {
    const payload: { [key: string]: string } = {};
    const input = event?.target?.value;
    payload[inputField] = input;
    setIsPreview(false);
    editDispatch({
      type: 'INPUT_CHANGE',
      payload: payload,
    });
  };

  const { Option } = Select;

  const intl = useIntl();

  const createNewAction = () => {
    const payload = {
      application: 'PHARMACY',
      emails: '',
      event: '',
      header: editState.headerName.trim(),
      optional_fields: optionalFields.length > 0 ? optionalFields : null,
      required_fields: requiredFields.length > 0 ? requiredFields : null,
      response: editState.responseText,
      screen_text: editState.screenText,
      ticket: editState.ticketPriority,
      type: editState.type,
    };
    const [error, data] = validateData(payload);
    if (error === false && data !== null) {
      dispatch(addNewNodeAction(accessToken, data));
    }
  };

  /*
   Function for validate data in form
  */

  const validateData = (payload: any): [boolean, null | object] => {
    let error = false;
    const submitData = { ...payload };
    if (submitData.type === 'INFO' || submitData.type === 'WEBVIEW') {
      submitData.optional_fields = null;
      submitData.required_fields = null;
      if (submitData.screen_text === '' || submitData.screen_text === null) {
        toast.error(ravenConstants.INFO_WEBVIEW_CHECK);
        error = true;
      }
    }
    if (submitData.type !== '' && submitData.response !== '') {
      if (
        submitData.type === 'FORM' ||
        submitData.type === 'INFO_WITH_WRITE_TO_US'
      ) {
        let fieldsLength = 0;
        if (submitData.optional_fields !== null) {
          fieldsLength += submitData.optional_fields.length;
        }
        if (submitData.required_fields !== null) {
          fieldsLength += submitData.required_fields.length;
        }
        if (fieldsLength === 0) {
          toast.error(ravenConstants.FIELDS_LENGTH_CHECK);
          error = true;
        }
        if (
          submitData.required_fields &&
          submitData.required_fields.indexOf('skus') > -1 &&
          fieldsLength !== 3
        ) {
          toast.error(ravenConstants.SKU_TYPE_CHECK);
          error = true;
        } else if (
          ((submitData.required_fields &&
            submitData.required_fields.indexOf('attachments') > -1) ||
            (submitData.optional_fields &&
              submitData.optional_fields.indexOf('attachments') > -1)) &&
          fieldsLength !== 3
        ) {
          toast.error(ravenConstants.ATTACHMENT_TYPE_CHECK);
          error = true;
        }
      }
    } else {
      toast.error(ravenConstants.NAME_TYPE_CHECK);
      error = true;
    }
    return [error, submitData];
  };

  /*
   Function for Optional field update
  */

  const optionalFieldUpdate = (fieldValue: string) => {
    if (optionalFields.includes(fieldValue)) {
      const newArr = [...optionalFields];
      const index = newArr.indexOf(fieldValue);
      newArr.splice(index, 1);
      setOptionalFields(newArr);
    } else {
      setOptionalFields([fieldValue, ...optionalFields]);
    }
  };

  /*
   Function for required field update
  */

  const requiredFieldUpdate = (fieldValue: string) => {
    if (requiredFields.includes(fieldValue)) {
      const newArr = [...requiredFields];
      const index = newArr.indexOf(fieldValue);
      newArr.splice(index, 1);
      setRequiredFields(newArr);
    } else {
      setRequiredFields([fieldValue, ...requiredFields]);
    }
  };

  const { TextArea } = Input;
  return (
    <Fragment>
      <div className="raven-wrapper">
        <div className=" mx-auto wallet-wrapper raven-wrap  shadow-md">
          <h2 className="text-lg raven-form-title font-semibold text-center py-1">
            <FormattedMessage id="create_action_form" />
          </h2>
          <div className=" py-2 px-3 flex-col flex">
            <div className="flex mb-3  gap-4">
              <div className="flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="node_action_header" />
                  <span className="px-1">:</span>
                </span>
                <span>
                  <Input
                    className={
                      'custom-input  focus:ring-blue-500 focus:outline-none appearance-none w-full text-sm leading-6  rounded-md py-1 px-1.5  shadow-sm'
                    }
                    placeholder={intl.formatMessage({ id: 'header_name' })}
                    type="text"
                    data-testid="headerName"
                    value={editState.headerName}
                    onChange={(e) => updateInputField(e, 'headerName')}
                  />
                  <div className="no-character">
                    {editState?.headerName?.trim().length > 0 && (
                      <span className="character">
                        <span className="text-sm font-semibold">
                          <FormattedMessage id="no_of_characters" />
                        </span>{' '}
                        <span className="px-1">:</span>
                        {editState?.headerName.trim().length}
                      </span>
                    )}
                  </div>
                </span>
              </div>
            </div>
            <div className="flex mb-3 gap-4">
              <div className="input-label flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="type_details" />
                  <span className="px-1">:</span>
                </span>
                <span>
                  <Select
                    placeholder={intl.formatMessage({ id: 'select_type' })}
                    className="w-full"
                    value={editState.type}
                    onChange={(value: string) => {
                      setIsPreview(false);
                      editDispatch({
                        type: 'INPUT_CHANGE',
                        payload: { type: value },
                      });
                      // Set Ticket Priority To None
                      editDispatch({
                        type: 'INPUT_CHANGE',
                        payload: { ticketPriority: '' },
                      });
                    }}
                  >
                    {ravenMetaInfo &&
                      ravenMetaInfo?.type?.map((type: any) => (
                        <Option key={type.id} value={type.type}>
                          {type.type}
                        </Option>
                      ))}
                  </Select>
                </span>
              </div>
              <div className="flex-1">
                <span className="text-sm font-semibold ">
                  <FormattedMessage id="ticket_priorty" />
                  <span className="px-1">:</span>
                </span>
                <span>
                  <Select
                    placeholder={intl.formatMessage({
                      id: 'select_ticket_priority',
                    })}
                    className="w-full"
                    value={editState.ticketPriority}
                    onChange={(value: string) => {
                      setIsPreview(false);
                      editDispatch({
                        type: 'INPUT_CHANGE',
                        payload: { ticketPriority: value },
                      });
                    }}
                  >
                    <Option key={0} value={''}>
                      None Priority
                    </Option>
                    {ticketPriorityObjects &&
                      ticketPriorityObjects?.map((obj) => (
                        <Option key={obj?.id} value={obj?.id}>
                          {obj?.status + '_' + obj?.priority}
                        </Option>
                      ))}
                  </Select>
                </span>
              </div>
            </div>
            <div className="flex mb-3 flex-col mt-3 gap-4">
              <div className="input-label flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="optional_fields" />
                  <span className="px-1">:</span>
                  {ravenMetaInfo &&
                    ravenMetaInfo?.fields?.map((nodeActionFields: any) => (
                      <CheckBoxFields
                        key={nodeActionFields.id}
                        changeFunction={optionalFieldUpdate}
                        name={'optional_fields'}
                        id={nodeActionFields.id}
                        fields={nodeActionFields.fields}
                        checkedField={optionalFields}
                        disabledField={requiredFields}
                        isFieldChecked={false}
                        skusCheck={
                          nodeActionFields.fields === 'skus' ? true : false
                        }
                      />
                    ))}
                </span>
                <span></span>
              </div>
              <div className="flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="required_fields" />
                  <span className="px-1">:</span>
                  {ravenMetaInfo &&
                    ravenMetaInfo?.fields?.map((nodeActionFields: any) => (
                      <CheckBoxFields
                        key={nodeActionFields.id}
                        changeFunction={requiredFieldUpdate}
                        name={'required_fields'}
                        id={nodeActionFields.id}
                        fields={nodeActionFields.fields}
                        checkedField={requiredFields}
                        disabledField={optionalFields}
                        isFieldChecked={false}
                        skusCheck={false}
                      />
                    ))}
                </span>
                <span></span>
              </div>
            </div>
            <div className="flex mb-3 flex-col gap-4">
              <div className="input-label flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="screen_text" />
                  <span className="px-1">:</span>
                </span>
                <span>
                  <TextArea
                    rows={4}
                    placeholder={`<b>Thank you!</b><br>We have received your message and will be in touch over your registered email id as soon as possible.`}
                    value={editState.screenText}
                    onChange={(e) => updateInputField(e, 'screenText')}
                  />
                  <div className="no-character">
                    {editState?.screenText?.length > 0 && (
                      <span className="character">
                        <span className="text-sm font-semibold">
                          <FormattedMessage id="no_of_characters" />
                        </span>{' '}
                        <span className="px-1">:</span>
                        {editState?.screenText?.length}
                      </span>
                    )}
                    {isPreview && editState.screenText.length > 0 && (
                      <div
                        className="screen-test"
                        dangerouslySetInnerHTML={{
                          __html: editState.screenText,
                        }}
                      />
                    )}
                  </div>
                </span>
              </div>
              <div className="flex-1">
                <span className="text-sm font-semibold">
                  <FormattedMessage id="response_text" />
                  <span className="px-1">:</span>
                </span>
                <span>
                  <TextArea
                    rows={4}
                    placeholder={`<b>Thank You!</b><br>We have received your message and will be in touch over your registered email id as soon as possible.`}
                    value={editState.responseText}
                    onChange={(e) => updateInputField(e, 'responseText')}
                  />
                  <div className="no-character">
                    {editState?.responseText?.length > 0 && (
                      <span className="character">
                        <span className="text-sm font-semibold">
                          <FormattedMessage id="no_of_characters" />
                        </span>{' '}
                        <span className="px-1">:</span>{' '}
                        {editState?.responseText?.length}
                      </span>
                    )}
                  </div>
                  {isPreview && editState.responseText.length > 0 && (
                    <div
                      className="screen-test"
                      dangerouslySetInnerHTML={{
                        __html: editState.responseText,
                      }}
                    />
                  )}
                </span>
              </div>
            </div>
            <div className="flex justify-center items-center">
              {!isPreview ? (
                <Button className="" onClick={() => setIsPreview(true)}>
                  <FormattedMessage id="preview" />
                </Button>
              ) : (
                <Button className="" onClick={() => createNewAction()}>
                  <FormattedMessage id="submit" />
                </Button>
              )}
            </div>
          </div>
        </div>
      </div>
    </Fragment>
  );
};
export default CreateAction;
