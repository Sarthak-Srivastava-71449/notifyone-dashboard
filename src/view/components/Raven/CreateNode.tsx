import { Input, Select } from 'antd';
import React, { Fragment, useEffect, useState } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { Button } from '../Button';
import { RootState } from 'src/store';
import { fetchRavenNodeActions } from 'src/store/actions/ravenActionNodeEvents';
import { fetchRavenNodes } from 'src/store/actions/ravenNodeEvents';
import { createNodeEvent } from 'src/store/actions/ravenRootNodeEvents';
import {
  getCreateNodeSuccess,
  getNewNodeOptions,
  getNodeActions,
} from 'src/store/selectors/createNode';
import { getAccessToken } from 'src/store/selectors/accessToken';
import './Raven.css';

const CreateNode: React.FC = () => {
  const [nodeName, setNodeName] = useState('');
  const [selectedNodeType, setSelectedNodeType] = useState('Select Node Type');
  const [selectedNode, setSelectedNode] = useState('');

  /*
   Redux Selectors for fetching data
 */

  const accessToken = useSelector((state: RootState) => getAccessToken(state));
  const createNodeSuccess = useSelector((state: RootState) =>
    getCreateNodeSuccess(state)
  );
  const newNodeOptions = useSelector((state: RootState) =>
    getNewNodeOptions(state)
  );
  const nodeActions = useSelector((state: RootState) => getNodeActions(state));

  const navigate = useNavigate();

  useEffect(() => {
    fetchData();
    if (createNodeSuccess === true) {
      navigate('/communication/raven', { replace: true });
    }
  }, [createNodeSuccess]);

  const { Option } = Select;
  const dispatch = useDispatch();
  const intl = useIntl();

  // fetch all the updated data
  const fetchData = () => {
    dispatch(fetchRavenNodes(accessToken, 200, 0));
    dispatch(fetchRavenNodeActions(accessToken, 200, 0));
  };

  useEffect(() => {
    let string = '';
    if (selectedNodeType === 'Sub Question') {
      string = 'Select Node';
    } else if (selectedNodeType === 'Node Action') {
      string = 'Select Node Action';
    } else {
      string = '';
    }
    setSelectedNode(string);
  }, [selectedNodeType]);

  // handler for node selection dropdown
  const handleChangeNode = (selectedNodeIdValue: string) => {
    setSelectedNode(selectedNodeIdValue);
  };

  // handler to create node submit button
  const createNodeHandler = () => {
    const payload = {
      application: 'PHARMACY',
      nodeName: nodeName.trim(),
      selectedNode: '',
      selectedNodeAction: '',
      selectedNodeType: selectedNodeType,
    };
    if (!parseInt(selectedNode)) {
      // Check if any Node ID is selected or not
      return;
    }
    if (selectedNodeType === 'Sub Question') {
      payload.selectedNode = selectedNode;
    } else if (selectedNodeType === 'Node Action') {
      payload.selectedNodeAction = selectedNode;
    }
    dispatch(createNodeEvent(accessToken, payload));
    setNodeName('');
    setSelectedNodeType('Select Node Type');
    setSelectedNode('');
  };

  return (
    <Fragment>
      <div className="mx-auto m-6 raven-wrap wallet-wrapper shadow-md">
        <h2 className="text-lg raven-form-title font-semibold text-center py-1">
          <FormattedMessage id="create_node_form" />
        </h2>
        <div className="px-3 py-2 flex-col flex">
          <div className="flex mb-3  gap-4">
            <div className="input-label flex-1">
              <h3
                data-testid="node_details_data"
                className="text-lg rounded-md  bg-blue-400   text-white px-1 font-semibold py-1"
              >
                <FormattedMessage id="node_details" />
              </h3>
              <div className="py-3">
                <span className="text-sm px-1 font-semibold">
                  <FormattedMessage id="enter_node_name" />
                  <span className="px-1">:</span>
                </span>
                <div>
                  <Input
                    className={
                      'w-full text-sm leading-6  rounded-md py-1  shadow-sm'
                    }
                    placeholder={intl?.formatMessage({
                      id: 'enter_node_name',
                    })}
                    type="text"
                    data-testid="enter_node_name"
                    onChange={(e) => setNodeName(e.target.value)}
                  />
                </div>
              </div>
            </div>
          </div>
          <div className="flex mb-3 gap-4">
            <div className="input-label flex-1">
              <h3 className="text-lg rounded-md  bg-blue-400   text-white px-1 font-semibold py-1">
                <FormattedMessage id="node_type" />
              </h3>
              <div className="py-3">
                <span className="text-sm px-1  font-semibold">
                  <FormattedMessage id="select_node_type" />
                  <span className="px-1">:</span>
                </span>
                <div>
                  <Select
                    placeholder={intl?.formatMessage({
                      id: 'select_node_type',
                    })}
                    value={selectedNodeType}
                    className="w-full text-sm leading-6  rounded-md shadow-sm"
                    onChange={(nodeType: string) =>
                      setSelectedNodeType(nodeType)
                    }
                  >
                    <Option value="Select Node Type">
                      <FormattedMessage id="select_node_type" />
                    </Option>
                    <Option value="Sub Question">
                      <FormattedMessage id="sub_question" />
                    </Option>
                    <Option value="Node Action">
                      {' '}
                      <FormattedMessage id="node_action" />
                    </Option>
                  </Select>
                </div>
              </div>
            </div>
            {selectedNodeType !== 'Select Node Type' && (
              <div className="flex-1">
                <h3 className="text-lg rounded-md  bg-blue-400   text-white px-1 font-semibold py-1">
                  {selectedNodeType === 'Sub Question' && (
                    <FormattedMessage id="node_details" />
                  )}
                  {selectedNodeType === 'Node Action' && (
                    <FormattedMessage id="node_action_details" />
                  )}
                </h3>
                <div className="py-3">
                  <span className="text-sm px-1 font-semibold">
                    {selectedNodeType === 'Sub Question' && (
                      <FormattedMessage id="select_node" />
                    )}
                    {selectedNodeType === 'Node Action' && (
                      <FormattedMessage id="select_node_action" />
                    )}
                    <span className="px-1">:</span>
                  </span>

                  <div>
                    <Select
                      placeholder={
                        selectedNodeType === 'Sub Question'
                          ? intl.formatMessage({
                              id: 'select_node',
                            })
                          : selectedNodeType === 'Node Action'
                          ? intl?.formatMessage({
                              id: 'select_node_action',
                            })
                          : ''
                      }
                      value={selectedNode}
                      className="w-full text-sm leading-6  rounded-md shadow-sm"
                      onChange={handleChangeNode}
                    >
                      {selectedNodeType === 'Sub Question' &&
                        newNodeOptions?.length > 0 &&
                        newNodeOptions.map((item: any) => (
                          <Option
                            key={item.id}
                            value={item?.id}
                          >{`${item?.name}`}</Option>
                        ))}
                      {selectedNodeType === 'Node Action' &&
                        nodeActions?.length > 0 &&
                        nodeActions.map((item: any) => (
                          <Option
                            key={item.id}
                            value={item?.id}
                          >{`${item?.header}`}</Option>
                        ))}
                    </Select>
                  </div>
                </div>
              </div>
            )}
          </div>
          <div className="flex justify-center items-center">
            <Button
              data-testid="node_Submit_button"
              className=""
              onClick={createNodeHandler}
              disabled={!parseInt(selectedNode) || nodeName?.trim().length < 1}
            >
              <FormattedMessage id="submit" />
            </Button>
          </div>
        </div>
      </div>
    </Fragment>
  );
};
export default React?.memo(CreateNode);
