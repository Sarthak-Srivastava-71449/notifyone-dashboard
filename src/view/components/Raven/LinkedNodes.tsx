import React from 'react';
import './Raven.css';
interface LinkedNodeProps {
  nodes: Array<any>;
}

const LinkedNodes: React.FC<LinkedNodeProps> = (props: LinkedNodeProps) => {
  return (
    <div className="container">
      <div className="popup">
        <div className="wrapper">
          <div className="card">
            <div className="header">Linked of Node IDs</div>
            <div className="row">
              <div className="col-xs-push-6 col-md-push-8">
                {props.nodes.map((linkedNodeDetail) => (
                  <div key={linkedNodeDetail.id} className="text-padding">
                    <span className="linknode-details-id px-1 font-semibold">
                      {linkedNodeDetail.id}.
                    </span>{' '}
                    &nbsp;
                    {linkedNodeDetail.name}
                  </div>
                ))}
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
export default LinkedNodes;
