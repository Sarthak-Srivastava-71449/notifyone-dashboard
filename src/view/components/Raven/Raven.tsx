import { Collapse } from 'antd';
import React, { useEffect } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { Spinner } from '../Spinner';
import { RootState } from 'src/store';
import { fetchRavenNodeActions } from 'src/store/actions/ravenActionNodeEvents';
import {
  fetchRavenInactiveNodeLinks,
  fetchRavenNodes,
} from 'src/store/actions/ravenNodeEvents';
import {
  fetchRavenMetaData,
  fetchRavenRootNodes,
  fetchRavenTickets,
} from 'src/store/actions/ravenRootNodeEvents';
import { getAccessToken } from 'src/store/selectors/accessToken';
import {
  getAddNodeSuccess,
  getChangeActionSuccess,
  getLoading,
  getRootNodes,
  getToggleUpdateSuccess,
} from 'src/store/selectors/raven';
import './Raven.css';
import SubNode from './SubNode';

const Raven: React.FC = () => {
  const dispatch = useDispatch();

  /*
  Redux Selectors for fetching data
*/
  const accessToken = useSelector((state: RootState) => getAccessToken(state));
  const shouldShowSpinner = useSelector((state: RootState) =>
    getLoading(state)
  );
  const rootNodes = useSelector((state: RootState) => getRootNodes(state));

  const addNodeSuccess = useSelector((state: RootState) =>
    getAddNodeSuccess(state)
  );
  const changeActionSuccess = useSelector((state: RootState) =>
    getChangeActionSuccess(state)
  );
  const toggleUpdateSuccess = useSelector((state: RootState) =>
    getToggleUpdateSuccess(state)
  );
  const updateRankSuccess = useSelector(
    (state: RootState) => state?.ravenRootEventsReducer?.updateRankSuccess
  );
  useEffect(() => {
    dispatch(fetchRavenRootNodes(accessToken, 200, 0));
    dispatch(fetchRavenNodeActions(accessToken, 200, 0));
    dispatch(fetchRavenNodes(accessToken, 200, 0));
    dispatch(fetchRavenInactiveNodeLinks(accessToken));
    dispatch(fetchRavenMetaData(accessToken));
    dispatch(fetchRavenTickets(accessToken));
  }, []);

  useEffect(() => {
    if (
      changeActionSuccess === true ||
      addNodeSuccess === true ||
      toggleUpdateSuccess === true ||
      updateRankSuccess === true
    ) {
      dispatch(fetchRavenRootNodes(accessToken, 200, 0));
      dispatch(fetchRavenInactiveNodeLinks(accessToken));
    }
  }, [
    changeActionSuccess,
    addNodeSuccess,
    toggleUpdateSuccess,
    updateRankSuccess,
  ]);

  const intl = useIntl();
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onChange = () => {};

  return shouldShowSpinner ? (
    <Spinner loading={shouldShowSpinner} />
  ) : (
    <div className="raven-wrapper">
      <h2 className="text-lg raven-form-title font-semibold text-center py-1">
        <FormattedMessage id="root_node" />
      </h2>
      {rootNodes?.length > 0 &&
        rootNodes.map((rootNode: any) => (
          <Collapse
            className=""
            // onChange={onChange}
            accordion
            key={rootNode.node_details.id}
          >
            <SubNode nodeSubQuestion={rootNode.node_details} rootNode={true} />
          </Collapse>
        ))}
    </div>
  );
};

export default Raven;
