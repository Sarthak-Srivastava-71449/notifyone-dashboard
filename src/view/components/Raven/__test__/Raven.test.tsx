import React from 'react';
import { withProvider } from '../../../../common/renderWithProvider';
import '@testing-library/jest-dom/extend-expect';
import {
  render,
  cleanup,
  waitFor,
  getByTestId,
  fireEvent,
  screen,
} from '@testing-library/react';
import Raven from '../Raven';

jest.mock('src/store/actions/ravenNodeEvents');

describe('Raven Test-Suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterEach(cleanup);
  test('Match Raven Snapshot', async () => {
    const { asFragment } = render(withProvider(<Raven />));
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders the landing page', () => {
    render(withProvider(<Raven />));
  });
});
