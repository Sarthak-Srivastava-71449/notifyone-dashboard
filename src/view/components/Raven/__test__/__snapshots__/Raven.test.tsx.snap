// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raven Test-Suite Match Raven Snapshot 1`] = `
<DocumentFragment>
  <div
    class="raven-wrapper"
  >
    <h2
      class="text-lg raven-form-title font-semibold text-center py-1"
    >
      Root Node
    </h2>
  </div>
</DocumentFragment>
`;
